yt/yt/server/query_tracker/handler_base.cpp

```diff
@@ -79,33 +79,54 @@ TSharedRef TruncateRowset(TSharedRange<NTableClient::TUnversionedRow>& rows)
-    newRecord.DataStatistics = ConvertToYsonString(dataStatistics);
-    newRecord.IsTruncated = wireSchemaAndSchemafulRowset.IsTruncated;
-    if (rowset.Size() <= MaxStringValueLength) {
-        // Fast path. Copy full rowset.
-        newRecord.Rowset = TString(rowset.ToStringBuf());
-    } else {
-        // Slow path. Truncate rowset.
+    try {
+        YT_LOG_DEBUG("Processing wire rowset");
+        // Process schema.
+        YT_LOG_DEBUG("Reading schema");
+        TWireProtocolOptions readerOptions;
+        readerOptions.MaxStringValueLength = ValueLengthLimit;
+        readerOptions.MaxAnyValueLength = ValueLengthLimit;
+        readerOptions.MaxCompositeValueLength = ValueLengthLimit;
+        auto reader = CreateWireProtocolReader(wireSchemaAndSchemafulRowset.Rowset, TRowBufferPtr(), readerOptions);
+        auto schema = reader->ReadTableSchema();
+        auto schemaNode = ConvertToNode(schema);
+        // Values in tables cannot have top-level attributes, but we do not need them anyway.
+        schemaNode->MutableAttributes()->Clear();
+        newRecord.Schema = ConvertToYsonString(schemaNode);
+
+        // Process schemaful rowset.
+        YT_LOG_DEBUG("Reading schemaful rowset");
+        auto rowset = reader->Slice(reader->GetCurrent(), reader->GetEnd());
+        auto schemaData = IWireProtocolReader::GetSchemaData(schema);
+        auto rows = reader->ReadSchemafulRowset(schemaData, /*captureValues*/ false);
+        TDataStatistics dataStatistics;
+        dataStatistics.set_row_count(rows.size());
+        dataStatistics.set_data_weight(GetDataWeight(rows));
+        newRecord.DataStatistics = ConvertToYsonString(dataStatistics);
+        newRecord.IsTruncated = wireSchemaAndSchemafulRowset.IsTruncated;
+        if (rowset.Size() <= MaxStringValueLength) {
+            // Fast path. Copy full rowset.
+            YT_LOG_DEBUG("Copying full rowset of size %v", rowset.Size());
+            newRecord.Error = TError();
+            newRecord.Rowset = TString(rowset.ToStringBuf());
+        } else {
+            // Slow path. Truncate rowset.
+            YT_LOG_DEBUG("Truncating rowset of size %v", rowset.Size());
+            newRecord.IsTruncated = true;
+            auto truncatedRowset = TruncateRowset(rows);
+            YT_LOG_DEBUG("Copying truncated rowset of size %v", truncatedRowset.Size());
+            newRecord.Error = TError();
+            newRecord.Rowset = TString(truncatedRowset.ToStringBuf());
+        }
+    } catch (const std::exception& ex) {
+        YT_LOG_WARNING(ex, "Failed to process rowset");
+        newRecord.Schema = ConvertToYsonString(TString());
+        newRecord.DataStatistics = ConvertToYsonString(TDataStatistics());
         newRecord.IsTruncated = true;
-        auto truncatedRowset = TruncateRowset(rows);
-        newRecord.Rowset = TString(truncatedRowset.ToStringBuf());
+        newRecord.Error = TError("Failed to save rowset") << ex;
```

Comment:
The errors might be different though, it could be easy to miss other problems this way, because clients will not report them. I always vote for more categorization/introspection.